codemod: add 'replace:' flag to frontend search

rvantonder opened the pull request:
Summary:
- 'replace:' qualifier to search. Active when `localStorage.codemodExp=true;location.reload()` is set.
- Uses GQL API and SearchParams/Result. A replace operation can be seen a kind of "diff search". This has only a small effect on the current GQL schema. This is great, since the change is small and allows us to revisit alternative streaming or cursor-based streaming that's under discussion for search.

Example:

<img width="1265" alt="Screen Shot 2019-06-17 at 5 10 47 PM" src="https://user-images.githubusercontent.com/888624/59637020-e5b0de00-9122-11e9-98f8-393d8affaed8.png">



Test plan: 
- Sane match/replace/file filter patterns
- Tested manually: Gate codemod GQL queries with `localStorage.codemodExp=true;location.reload()`

sqs commented on cmd/frontend/internal/app/ui/router.go:
```diff
@@ -208,6 +210,7 @@ func initRouter() {
 	router.Get(routeRepoCompare).Handler(handler(serveBasicPageString("Compare - Sourcegraph")))
 	router.Get(routeRepoStats).Handler(handler(serveBasicPageString("Stats - Sourcegraph")))
 	router.Get(routeRepoGraph).Handler(handler(serveBasicPageString("Repository graph - Sourcegraph")))
+	router.Get(routeRepoGraph).Handler(handler(serveBasicPageString("Codemods - Sourcegraph")))
```
```suggestion
	router.Get(routeCodemods).Handler(handler(serveBasicPageString("Codemods - Sourcegraph")))
```

sqs commented:
Did you see my branch's code, eg at https://sourcegraph.com/github.com/sourcegraph/sourcegraph@codemod-ui/-/blob/cmd/frontend/graphqlbackend/codemod.go? All of the code I wrote related to codemod in Go, replacer service, etc., is 100% prototype and hacky, and it's fine to ignore. But it might be helpful.

rvantonder commented:
Yep, this is mostly your `codemod` code that I pulled out of that commit :) I'm fixing up some things and was going to ping your for comments after.

rvantonder commented:
In the interest of end-to-end codemod, my proposal is that we get this code in a state where we can actually merge it, to enable internal experimenting ([discussion](https://docs.google.com/document/d/14A0bfQ9Nef2JBmovlCrIRalTEc_U14K2roQjHKnOvzo/edit#heading=h.k9jig167fef9). 

The plan is to still write tests and some clean up, but besides that:

- We should gate the `replace:` codepath  initially. What do you think @sqs, maybe with a setting in the browser console? 
- Since this is mostly from your `codemod-ui` branch @sqs, what are the especially hacky parts that need attention? (Hopefully we'd mitigate the extent of hacks by hiding this codepath behind a flag)

rvantonder commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,261 @@
+		replacementText = replacementValues[0]
+	}
+
+	fileFilter, _ := args.Query.RegexpPatterns(query.FieldFile)
```
The `FieldFile` is a useful/essential way to choose the codemod language matcher and filter individual files. I'm abusing it a little here, because `comby` won't accept a regex pattern for matching files yet, and so filter for only alphanumeric characters. The other option is to use the `language` field, but that can't actually filter files. It doesn't need to happen in this PR.

sqs commented:
> We should gate the replace: codepath initially. What do you think @sqs, maybe with a setting in the browser console?

Yep - on the frontend I think I have `localStorage.codemodExp` as the feature flag: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@codemod-ui/-/blob/web/src/enterprise/codemod/index.ts#L4:38. It's fine to NOT gate it on the backend (just ensure that repo ACLs are respected). This means that anyone searching for a literal `replace:whatever` would get an error, but that's OK.

> Since this is mostly from your codemod-ui branch @sqs, what are the especially hacky parts that need attention? (Hopefully we'd mitigate the extent of hacks by hiding this codepath behind a flag)

Hacky parts in this PR from my code are:

- the search result React UI element that shows the diff and highlights the replacements
- the Go code that calls codemod is copy-pasta'd

rvantonder commented:
> Yep - on the frontend I think I have `localStorage.codemodExp` as the feature flag: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@codemod-ui/-/blob/web/src/enterprise/codemod/index.ts#L4:38. 

Cool. Looks like that gates additional UI views that I'm not putting in this PR. I've kept that file and gated this at the place where a GQL query is issued (see inline).

> Hacky parts in this PR from my code are:
> 
> * the search result React UI element that shows the diff and highlights the replacements

I've removed the ad-hoc highlighting for now. The neat structural highlighting is really appealing, but for simplicity I'm leaving it out in this PR as the standard diff view is adequate. Will reintroduce highlighting soon, as I have some ideas here and want to play with it a bit more.

> * the Go code that calls codemod is copy-pasta'd

I've cleaned things up generally, and I think this is in a better state now.

rvantonder commented on web/src/search/backend.tsx:
```diff
@@ -39,8 +40,13 @@ export function search(
-        switchMap(query =>
-            queryGraphQL(
+        switchMap(query => {
+            const codemodActive = USE_CODEMOD
```
Gating happens here.

rvantonder commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,274 @@
+			matchTemplates = append(matchTemplates, *v.String)
+		}
+		if v.Regexp != nil || v.Bool != nil {
+			return "", "", "", errors.New("This looks like a regex search pattern. Structural search is active because 'replace:' was specified. Please enclose your search string with quotes when using 'replace:'.")
```
I'm opting for strictly-non-regex search input rather than trying to massage regex input into a quoted search string. This was previously:
```
		if v.Regexp != nil {
			// HACK
			matchPatterns = append(matchPatterns, strings.Replace(v.Regexp.String(), "\\", "", -1))
		}
```

ijt replied:
I don't understand the reasoning here. If `replace:` was specified, wouldn't that mean we should treat all non-fields as if they are quoted?

ijt replied:
I just saw your example in the description for this PR. Makes sense.

rvantonder commented:
Finalllyyy ready for review. Adding @sqs since this is adapted from `codemod-ui` branch.

ijt commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,274 @@
+	if len(fileFilter) > 0 {
+		fileFilterText = fileFilter[0]
+		// only file names or files with extensions in the following characterset are allowed
+		var IsAlphanumericWithPeriod = regexp.MustCompile(`^[a-zA-Z_.]+$`).MatchString
```
I don't see the numeric part in `[a-zA-Z_.]`. Shouldn't it be `[a-zA-Z0-9_.]`?

rvantonder replied:
Yes, good eye.

ijt commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,274 @@
+	}()
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
```
Why cancel?

rvantonder replied:
So some of this code is coming from other search code, which similarly uses `defer cancel()`: https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/graphqlbackend/search_commits.go#L451. My understanding is it is clean up for contexts--if you have a better idea LMK (or let's find out).

ijt replied:
Okay, I think it's fine since it doesn't affect the passed in context.

ijt commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,274 @@
+		return nil, nil, err
+	}
+
+	tr, ctx := trace.New(ctx, "callCodemod", fmt.Sprintf("pattern: %+v, replace: %+v, fileFilter: %+v, numRepoRevs: %d", matchTemplate, rewriteTemplate, fileFilter, len(args.Repos)))
```
```suggestion
	title := fmt.Sprintf("pattern: %+v, replace: %+v, fileFilter: %+v, numRepoRevs: %d", matchTemplate, rewriteTemplate, fileFilter, len(args.Repos))
	tr, ctx := trace.New(ctx, "callCodemod", title)
```

ijt commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,274 @@
+	return matchTemplate, rewriteTemplate, fileFilterText, nil
+}
+
+func callCodemod(ctx context.Context, args *search.Args) ([]*searchResultResolver, *searchResultsCommon, error) {
```
It would help to have a comment describing what this function does.

ijt commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,274 @@
+			mu.Lock()
+			defer mu.Unlock()
+			if fatalErr := handleRepoSearchResult(common, repoRev, false, repoTimedOut, searchErr); fatalErr != nil {
+				err = errors.Wrapf(searchErr, "failed to call codemod %s", repoRev.String())
```
```suggestion
				err = errors.Wrapf(searchErr, "failed to call codemod %s", repoRev)
```

ijt replied:
`%s` calls `String()`.

rvantonder replied:
TIL.

ijt commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,274 @@
+		nethttp.ClientTrace(false))
+	defer ht.Finish()
+
+	resp, err := ctxhttp.Do(ctx, searchHTTPClient, req)
```
Neat. TIL!

rvantonder replied:
Uh, sqs wrote this :)

ijt commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,274 @@
+	// For performance, assume repo is cloned in gitserver and do not trigger a repo-updater lookup (this call fails if repo is not on gitserver).
+	commit, err := git.ResolveRevision(ctx, repoRevs.GitserverRepo(), nil, repoRevs.Revs[0].RevSpec, &git.ResolveRevisionOptions{NoEnsureRevision: true})
+	if err != nil {
+		return nil, errors.Wrap(err, "Codemod repo lookup failed: it's possible that the repo is not cloned in gitserver. Try force a repo update another way.")
```
The convention in Go errors is to not start them with a capital letter.

ijt replied:
https://github.com/golang/go/wiki/Errors

rvantonder replied:
💯 TIL

ijt commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,274 @@
+	return r.matches
+}
+
+func validateQuery(q *query.Query) (string, string, string, error) {
```
This would be a more readable if it returned a struct instead of three anonymous strings.

rvantonder replied:
Agree. I had to update this since anyway as more fields are needed in the query to include/exclude files.

ijt commented on cmd/frontend/graphqlbackend/search_results.go:
```diff
@@ -1094,6 +1139,10 @@ func (g *searchResultResolver) ToCommitSearchResult() (*commitSearchResultResolv
 	return g.diff, g.diff != nil
 }
 
+func (g *searchResultResolver) ToCodemodResult() (*codemodResultResolver, bool) {
```
Why `g` here instead of `r`?

rvantonder replied:
Seems to be in consistency with other functions:
`(g *searchResultResolver) ToRepository()`
`(g *searchResultResolver) ToCommitSearchResult()`
...

Not sure why `g` is chosen elsewhere.

slimsag commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,315 @@
+	)
+	for _, repoRev := range args.Repos {
+		wg.Add(1)
+		go func(repoRev search.RepositoryRevisions) {
```
use `goroutine.Go` here, like so:

```Go
repoRev := repoRev // shadow variable so it doesn't change while goroutine is running
goroutine.Go(func() {
    defer wg.Done()
    ...
})
```

This is done for all goroutines in the frontend so that one going down (like experimental codemod) doesn't take down the rest of the frontend (i.e. all of Sourcegraph) due to a bug.

slimsag commented on cmd/frontend/graphqlbackend/codemod.go:
```diff
@@ -0,0 +1,315 @@
+		nethttp.ClientTrace(false))
+	defer ht.Finish()
+
+	resp, err := ctxhttp.Do(ctx, searchHTTPClient, req)
```
```suggestion
    // TODO(RVT): Use a separate HTTP client here dedicated to codemod,
    // not doing so means codemod and searcher share the same HTTP limits
    // etc. which is fine for now but not if codemod goes in front of users.
	resp, err := ctxhttp.Do(ctx, searchHTTPClient, req)
```

rvantonder commented:
> LGTM, I checked and it appears to me ACLs are enforced properly / the same way it is for other search result types so security-wise LGTM as well.

Thanks! Yes should have mentioned, I did check auth; it passes through https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/frontend/db/repos.go#L177.
